src/util/process.cc

```diff
@@ -0,0 +1,83 @@
+/*
+ * This file is open source software, licensed to you under the terms
+ * of the Apache License, Version 2.0 (the "License").  See the NOTICE file
+ * distributed with this work for additional information regarding copyright
+ * ownership.  You may not use this file except in compliance with the License.
+ *
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/*
+ * Copyright (C) 2022 Kefu Chai ( tchaikov@gmail.com )
+ */
+
+#include <seastar/util/process.hh>
+#include <seastar/core/reactor.hh>
+#include <seastar/core/internal/buffer_allocator.hh>
+#include <seastar/core/polymorphic_temporary_buffer.hh>
+
+namespace seastar {
+
+class pipe_data_source_impl final : public data_source_impl {
+    static constexpr std::size_t buffer_size = 8192;
+    struct buffer_allocator : public internal::buffer_allocator {
+        temporary_buffer<char> allocate_buffer() override {
+            return make_temporary_buffer<char>(memory::malloc_allocator, buffer_size);
+        }
+    };
+public:
+    explicit pipe_data_source_impl(pollable_fd fd)
+        : _fd(std::move(fd)) {}
+    future<temporary_buffer<char>> get() override {
+        return _fd.read_some(&_ba);
+    }
+    future<> close() override {
+        _fd.close();
+        return make_ready_future();
+    }
+private:
+    pollable_fd _fd;
+    buffer_allocator _ba;
+};
+
+static auto make_pipe_input_stream(int fd) {
+    auto ds = std::make_unique<pipe_data_source_impl>(pollable_fd(file_desc::from_fd(fd)));
+    return input_stream<char>(data_source(std::move(ds)));
+}
+
+process::process(pid_t pid, int stdout, int stderr)
+    : _pid(pid),
+      _stdout(make_pipe_input_stream(stdout)),
+      _stderr(make_pipe_input_stream(stderr)) {}
+
+future<process::wait_status> process::wait() {
+    return engine().waitpid(_pid).then([] (int wstatus) -> wait_status {
+        if (WIFEXITED(wstatus)) {
+            return exit_status{WEXITSTATUS(wstatus)};
+        } else if (WIFSIGNALED(wstatus)) {
+            return term_sig{WTERMSIG(wstatus)};
+        } else {
+            return exit_status{-1};
+        }
+    });
+}
+
+future<process> spawn_process(std::string_view pathname,
```

Comment:
Please, keep it in `experimental` namespace so that we're not doomed to backward compatibility right at once

Reply:
sure. will do.